### PR TITLE
Implement batched serial tbsv

### PR DESCRIPTION
**Pull request by yasahi-hpc**
This PR implements [tbsv](https://oneapi-src.github.io/oneMKL/domains/blas/tbsv.html) function.

Following files are added:
1. `KokkosBatched_Tbsv_Serial_Impl.hpp`: Internal interfaces
2. `KokkosBatched_Tbsv_Serial_Internal.hpp`: Implementation details
3. `KokkosBatched_Tbsv.hpp`: APIs
4. `Test_Batched_SerialTbsv.hpp`: Unit tests for that

## Detailed description
 It solves the equation `Ax = b`.
Here, the matrix has the following shape.
- `A`: `(batch_count, lda, n)`  
   `n` by `n` unit or non-unit, upper or lower triangular band matrix with `(k+1)` diagonals.
- `x`: `(batch_count, n)`  
   Before entry, the incremented array x must contain the n element right-hand side vector `b`.

Example of a single batch of matrix A with `n = 10` and `k = 3`.
```bash
upper
1 1 1 1 0 0 0 0 0 0 
0 1 1 1 1 0 0 0 0 0 
0 0 1 1 1 1 0 0 0 0 
0 0 0 1 1 1 1 0 0 0 
0 0 0 0 1 1 1 1 0 0 
0 0 0 0 0 1 1 1 1 0 
0 0 0 0 0 0 1 1 1 1 
0 0 0 0 0 0 0 1 1 1 
0 0 0 0 0 0 0 0 1 1 
0 0 0 0 0 0 0 0 0 1 

lower
1 0 0 0 0 0 0 0 0 0 
1 1 0 0 0 0 0 0 0 0 
1 1 1 0 0 0 0 0 0 0 
1 1 1 1 0 0 0 0 0 0 
0 1 1 1 1 0 0 0 0 0 
0 0 1 1 1 1 0 0 0 0 
0 0 0 1 1 1 1 0 0 0 
0 0 0 0 1 1 1 1 0 0 
0 0 0 0 0 1 1 1 1 0 
0 0 0 0 0 0 1 1 1 1 

upper_banded
0 0 0 1 1 1 1 1 1 1 
0 0 1 1 1 1 1 1 1 1 
0 1 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 1 

lower_banded
1 1 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 0 
1 1 1 1 1 1 1 1 0 0 
1 1 1 1 1 1 1 0 0 0 
```

Parallelization would be made in the following manner. This is efficient only when 
A is given in `LayoutLeft` for GPUs and `LayoutRight` for CPUs (parallelized over batch direction).

```C++
Kokkos::parallel_for('tbsv', 
    Kokkos::RangePolicy<execution_space, ParamTagType> policy(0, n),
    [=](const ParamTagType &, const int k) {
        auto aa = Kokkos::subview(_a, k, Kokkos::ALL(), Kokkos::ALL());
        auto bb = Kokkos::subview(_b, k, Kokkos::ALL());

        KokkosBatched::SerialTbsv<
            typename ParamTagType::uplo, typename ParamTagType::trans,
            typename ParamTagType::diag, AlgoTagType>::invoke(aa, bb, k, incx);
    });
```

## Tests
1. Firstly, construct an Upper or Lower banded diagonal matrix `A`. Then, convert it to the banded storage `Ab`. Solving `Ax = b` with `trsv` and `(Ab)x = b` with `tbsv`. `tbsv` result is compared with `trsv` result.
2.  Simple and small analytical test, i.e. chose `A` and `b` and compute by hand `x`, then check that it matches with your routine.

## Tasks
- [x] Add a simple and small analytical test, i.e. chose A and b and compute by hand x, then check that it matches with your routine.
- [ ] Do some corner cases, what happens when you have an empty matrix? Zero upper/lower diagonals, zeros on the diagonal, etc... (testing zero on diagonals must be improved)
- [x] Generate random matrices for A and x, compute b with a matvec and then check that solving Ax=b returns the original x vector. (This is partially done. Tbsv is compared with Trsv)
- [x] Ideally if you have any assertions/runtime checks that can be done, you want to check that they are catching what they are supposed to catch
- [x] Finally you want the test to cover all backends for which the algorithms is supposed to run. (Serial, OpenMP, Thread, CUDA, HIP can be tested)

### CODE REVIEW

**Comment by brian-kelley** [comment]
@yasahi-hpc Thanks for the new feature, I will review it but first could you please rebase onto the develop branch?
Also we merge new PRs into develop branch, so I changed the target of this PR from master to develop.

File: batched/dense/unit_test/Test_Batched_SerialTbsv.hpp [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+                            AlgoTagType>(Ab, x0, k, 1)
+      .run();
+
+  // tbsv with incx == 2
+  Functor_BatchedSerialTbsv<DeviceType, View3DType, View2DType, ParamTagType,
+                            AlgoTagType>(Ab, x1, k, incx)
+      .run();
+
+  // Check x0 = x_ref
+  EXPECT_TRUE(allclose<execution_space>(x0, x_ref, 1.e-5, 1.e-12));
```
Comment (lucbv) [inline batched/dense/unit_test/Test_Batched_SerialTbsv.hpp:295]: This is a common mistake: you should not pick arbitrary tolerances here you should at least use something like `Kokkos::ArithTraits<ScalarType>::eps()` so it works for both `double` and `float`.
In practice you want to figure out how many operations are performed to compute each entry in x and use that number to estimate how many ulp you have accumulated, then multiply epsilon by that number. Finally that check should be done against a relative error so there is no scaling issues with the tolerance.

File: batched/dense/unit_test/Test_Batched_SerialTbsv.hpp [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+                                  typename ParamTagType::uplo>(ref, Ab, k,
+                                                               true);
+
+  // tbsv
+  Functor_BatchedSerialTbsv<DeviceType, View3DType, View2DType, ParamTagType,
+                            AlgoTagType>(Ab, x0, k, 1)
+      .run();
+
+  // tbsv with incx == 2
+  Functor_BatchedSerialTbsv<DeviceType, View3DType, View2DType, ParamTagType,
```
Comment (lucbv) [inline batched/dense/unit_test/Test_Batched_SerialTbsv.hpp:290]: Users should not need to use `incx`, that information usually is carried by the layout of the `Kokkos::View` x1

File: batched/dense/src/KokkosBatched_Tbsv.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/// No nested parallel_for is used inside of the function.
+///
+
+template <typename ArgUplo, typename ArgTrans, typename ArgDiag,
+          typename ArgAlgo>
+struct SerialTbsv {
+  template <typename AViewType, typename XViewType>
+  KOKKOS_INLINE_FUNCTION static int invoke(const AViewType &A,
+                                           const XViewType &X, const int k,
+                                           const int incx);
```
Comment (lucbv) [inline batched/dense/src/KokkosBatched_Tbsv.hpp:52]: Since `X` is a `Kokkos::View`, using a layout stride is the normal way to represent `incx`

File: batched/dense/impl/KokkosBatched_Tbsv_Serial_Impl.hpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+#include "KokkosBatched_Tbsv_Serial_Internal.hpp"
+
+namespace KokkosBatched {
+
+template <typename AViewType, typename XViewType>
+KOKKOS_INLINE_FUNCTION static int checkTbsvInput(
+    [[maybe_unused]] const AViewType &A, [[maybe_unused]] const XViewType &x,
+    [[maybe_unused]] const int k, [[maybe_unused]] const int incx) {
+#if (KOKKOSKERNELS_DEBUG_LEVEL > 0)
+  static_assert(Kokkos::is_view<AViewType>::value,
```
Comment (lucbv) [inline batched/dense/impl/KokkosBatched_Tbsv_Serial_Impl.hpp:32]: Static assert are a compile time feature, why do you have them protected behind `KOKKOSKERNELS_DEBUG_LEVEL`? These compile time features do not increase the execution of the compiler code, I would move the check out of the pre-processor guard

**Comment by kokkos-devops-admin** [comment]
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

File: batched/dense/unit_test/Test_Batched_SerialTbsv.hpp (hunk 1 above)
Comment (yasahi-hpc) [reply batched/dense/unit_test/Test_Batched_SerialTbsv.hpp:295]: Sure. I will do that.

File: batched/dense/unit_test/Test_Batched_SerialTbsv.hpp (hunk 2 above)
Comment (yasahi-hpc) [reply batched/dense/unit_test/Test_Batched_SerialTbsv.hpp:290]: I am not quite sure about this point. If I understand correctly, `incx` can be a negative number, so it cannot be computed from View extents only.

File: batched/dense/impl/KokkosBatched_Tbsv_Serial_Impl.hpp (hunk 4 above)
Comment (yasahi-hpc) [reply batched/dense/impl/KokkosBatched_Tbsv_Serial_Impl.hpp:32]: I fully agree with you. I will fix accordingly. I found static assertions are protected in [some function](https://github.com/kokkos/kokkos-kernels/blob/d1a91b8a1f3aa4a972d7ff198bce73ec3248f641/batched/dense/impl/KokkosBatched_HadamardProduct_Impl.hpp#L94-L109), so I implement this way for consistency.

**Comment by yasahi-hpc** [comment]
I will close this PR. I made a new one from develop branch. #2202

**Comment by lucbv** [comment]
Superseded by PR #2202
